### PR TITLE
Ensure processing of CHARMM residues and patches in deterministic order

### PR DESCRIPTION
**Pull request by jchodera**
This PR ensures that patches and residues are processed and compared in deterministic order---the same order as read in. 

This allows us to avoid issues like [this](https://github.com/choderalab/openmm-forcefields/pull/61#issuecomment-365117418) where patches are compared against random base residues for determining what should be added/deleted in a patch.

### CODE REVIEW

File: parmed/openmm/parameters.py [hunk 1]
```diff
@@ -592,6 +601,11 @@ def _write_omm_patches(self, xml_root, valid_residues_for_patch, write_apply_to_
             if (residue.tail is None) and (patched_residue.tail is not None):
                 etree.SubElement(patch_xml, 'AddExternalBond', atomName=patched_residue.tail.name)
 
+            # DEBUG
+            if (patch.name == 'CTER'):
+                print('residue: {}'.format(residue))
+                print('patched_residue: {}'.format(patched_residue))
+
```
Comment (swails) [inline parmed/openmm/parameters.py]: Remove the debug code before merging.

Comment (jchodera) [reply parmed/openmm/parameters.py]: Whoops! Thanks!

**Comment by jchodera** [comment]
@swails : All travis branches pass, except for [this one](https://travis-ci.org/ParmEd/ParmEd/jobs/341706182) which just timed out.

**Comment by jchodera** [comment]
This is ready to merge @swails. The failure is a travis timeout.

**Pull request merged**